Rx/v2/src/rxcpp/operators/rx-timeout.hpp

```diff
@@ -0,0 +1,218 @@
+
+        struct timeout_subscriber_values : public timeout_values
+        {
+            timeout_subscriber_values(composite_subscription cs, dest_type d, timeout_values v, coordinator_type c)
+                : timeout_values(v)
+                , cs(std::move(cs))
+                , dest(std::move(d))
+                , coordinator(std::move(c))
+                , worker(coordinator.get_worker())
+                , index(0)
+            {
+            }
+
+            composite_subscription cs;
+            dest_type dest;
+            coordinator_type coordinator;
+            rxsc::worker worker;
+            mutable std::size_t index;
+        };
+        typedef std::shared_ptr<timeout_subscriber_values> state_type;
+        state_type state;
+
+        timeout_observer(composite_subscription cs, dest_type d, timeout_values v, coordinator_type c)
+            : state(std::make_shared<timeout_subscriber_values>(timeout_subscriber_values(std::move(cs), std::move(d), v, std::move(c))))
+        {
+            auto localState = state;
+
+            auto disposer = [=](const rxsc::schedulable&){
+                localState->cs.unsubscribe();
+                localState->dest.unsubscribe();
+                localState->worker.unsubscribe();
+            };
+            auto selectedDisposer = on_exception(
+                [&](){ return localState->coordinator.act(disposer); },
+                localState->dest);
+            if (selectedDisposer.empty()) {
+                return;
+            }
+
+            localState->dest.add([=](){
+                localState->worker.schedule(selectedDisposer.get());
+            });
+            localState->cs.add([=](){
+                localState->worker.schedule(selectedDisposer.get());
+            });
+        }
+
+        static std::function<void(const rxsc::schedulable&)> produce_timeout(std::size_t id, state_type state) {
+            auto produce = [id, state](const rxsc::schedulable&) {
+                if(id != state->index)
+                    return;
+
+                state->dest.on_error(std::make_exception_ptr(rxcpp::timeout_error("timeout has occurred")));
+            };
+
+            auto selectedProduce = on_exception(
+                    [&](){ return state->coordinator.act(produce); },
+                    state->dest);
+            if (selectedProduce.empty()) {
+                return std::function<void(const rxsc::schedulable&)>();
+            }
+
+            return std::function<void(const rxsc::schedulable&)>(selectedProduce.get());
+        }
+
+        void on_next(T v) const {
+            auto localState = state;
+            auto work = [v, localState](const rxsc::schedulable&) {
+                auto new_id = ++localState->index;
+                auto produce_time = localState->worker.now() + localState->period;
+
+                localState->dest.on_next(v);
+                localState->worker.schedule(produce_time, produce_timeout(new_id, localState));
```

Comment:
This looks correct. 

FYI - if the returned subscription was stored instead of the index then this could just unsubscribe from the previous timeout and replace with the new subscription.